README
Rack::Cache
===========

Rack::Cache is suitable as a quick drop-in component to enable HTTP caching for
Rack-based applications that produce freshness (Expires, Cache-Control) and/or
validation (Last-Modified, ETag) information:

  * Standards-based (RFC 2616)
  * Freshness/expiration based caching
  * Validation (If-Modified-Since / If-None-Match)
  * Vary support
  * Cache-Control: public, private, max-age, s-maxage, must-revalidate,
    and proxy-revalidate.
  * Portable: 100% Ruby / works with any Rack-enabled framework
  * Disk, memcached, and heap memory storage backends

For more information about Rack::Cache features and usage, see:

http://rtomayko.github.com/rack-cache/

Rack::Cache is not overly optimized for performance. The main goal of the
project is to provide a portable, easy-to-configure, and standards-based
caching solution for small to medium sized deployments. More sophisticated /
high-performance caching systems (e.g., Varnish, Squid, httpd/mod-cache) may be
more appropriate for large deployments with significant throughput requirements.

Installation
------------

From Gem:

    $ sudo gem install rack-cache

With a local working copy:

    $ git clone git://github.com/rtomayko/rack-cache.git
    $ rake package && sudo rake install

Basic Usage
-----------

Rack::Cache is implemented as a piece of Rack middleware and can be used with
any Rack-based application. If your application includes a rackup (`.ru`) file
or uses Rack::Builder to construct the application pipeline, simply require
and use as follows:

    require 'rack/cache'

    use Rack::Cache,
      :metastore   => 'file:/var/cache/rack/meta',
      :entitystore => 'file:/var/cache/rack/body',
      :verbose     => true

    run app

Assuming you've designed your backend application to take advantage of HTTP's
caching features, no further code or configuration is required for basic
caching.

Using with Rails
----------------

Add this to your `config/environment.rb`:

   config.middleware.use Rack::Cache,
       :verbose => true,
       :metastore   => 'file:/var/cache/rack/meta',
       :entitystore => 'file:/var/cache/rack/body'

You should now see `Rack::Cache` listed in the middleware pipeline:

    rake middleware

See the following for more information:

    http://snippets.aktagon.com/snippets/302-how-to-setup-and-use-rack-cache-with-rails

Using with Dalli
----------------

Dalli is a high performance memcached client for Ruby.
More information at: https://github.com/mperham/dalli

    require 'dalli'
    require 'rack/cache'

    use Rack::Cache,
      :verbose => true,
      :metastore   => "memcached://localhost:11211/meta",
      :entitystore => "memcached://localhost:11211/body"

    run app

Links
-----

Documentation:
    http://rtomayko.github.com/rack-cache/

GitHub:
    http://github.com/rtomayko/rack-cache/

License
-------

Copyright (c) 2008 Ryan Tomayko <http://tomayko.com/about>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to
deal in the Software without restriction, including without limitation the
rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
sell copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGES
## 1.13.0

 * Support lowercase Vary and Age headers

## 1.12.0

 * Add a fault_tolerant flag to fail-over to stale cache

## 1.11.1

 * when ignoring parts of the query, remove query in key when all params are ignored

## 1.11.0

 * Add a proc to allow ignoring parts of the query string in the key

## 1.10.0

  * Pass Options To Underlying Storage Driver
  * bump required ruby version to 2.3

## 1.9.0

  * make purge not raise when not implemented

## 1.8.0

  * Meta stores will purge keys when no entity store entries are found

## 1.7.2

  * Fix key generation for requests with no query strings

## 1.7.0

  * Meta stores now receive a third ttl argument to write when use_native_ttl is used.

## 1.6.1

  * Revert 'No longer read responses from cache when we already have them'

## 1.6.0

  * Noop backend
  * No longer read responses from cache when we already have them
  * renamed files from entitystore -> entity_store (metastore/cachecontrol/appengine) and added warns for old ones

## 1.5.1

  * fix key generation for query strings that include encoded equals

## 1.5.0

  * only catch StandardError and not Exception

## 1.4.3

  * After overriding the REQUEST_METHOD, store the original request method in "rack.methodoverride.original_method"

## 1.4.1

  * Ignore invalid Expires date as per RFC

## 1.4.0

  * Not invalidating the cache for preflight CORS request

## 1.3.1 / October 2015

  * Support Ruby 1.9

## 1.3 / Octorber 2015

  * Ruby 2.0 only

  * Gracefully degrade when cache store goes offline

  * allow_reload/revalidate is not enabled by default

  * Make Rack::Cache multithread friendly

## 1.2 / March 2012

  * Fix a cookie leak vulnerability effecting large numbers of Rails 3.x installs:
    https://github.com/rtomayko/rack-cache/pull/52

  * Never 304 on PUT or POST requests.

  * Misc bundler and test tooling fixes.

## 1.1 / September 2011

  * Allow (INM/IMS) validation requests through to backend on miss. Makes it
    possible to use validation for private / uncacheable responses. A number of
    people using Rails's stale?() helper reported that their validation logic was
    never kicking in.

  * Add rack env rack-cache.force-pass option to bypass rack-cache on
    per request basis

  * Fix an issue with memcache namespace not being set when using the
    :namespace option instead of :prefix_key.

  * Fix test failures due to MockResponse changes in recent Rack
    version (issue #34)

## 1.0.3 / August 2011

  * Fix bug passing options to memcached and dalli

  * Document cache_key

## 1.0.1 / April 2011

  * Added lib/rack-cache.rb to match package name for auto-requiring machinery.

  * Fixed a number of issues caused by Rack::Cache not closing the body received
    from the application. Rack::Lock and other middleware use body.close to
    signal the true end of request processing so failure to call this method
    can result in strange issues (e.g.,
    "ThreadError: deadlock; recursive locking")

  * Fixed a bug where Rack::Cache would blow up writing the rack env to the meta
    store when the env contained an all uppercase key whose value wasn't
    marshalable. Passenger and some other stuff write such keys apparently.

  * The test suite has moved from test-spec to bacon. This is a short term
    solution to the problem of not being able to run tests under Ruby 1.9.x.
    The test suite will be moved to basic Test::Unit style sometime in the
    future.

## 1.0 / December 2010

  * Rack::Cache is 1.0 and will now maintain semantic versioning <http://semver.org/>

  * Add Dalli memcache client support and removed support for the unmaintained
    memcache-client library. You will need to move your apps to Dalli before
    upgrading rack-cache to 1.0.

## 0.5.3 / September 2010

  * A matching If-Modified-Since is ignored if an If-None-Match is also provided
    and doesn't match. This is in line with RFC 2616.

  * Converts string status codes to integers before returns to workaround bad
    behaving rack middleware and apps.

  * Misc doc clean up.

## 0.5.2 / September 2009

  * Exceptions raised from the metastore are not fatal. This makes a lot of
    sense in most cases because its okay for the cache to be down - it
    shouldn't blow up your app.

## 0.5.1 / June 2009

  * Added support for memcached clusters and other advanced
    configuration provided by the memcache-client and memcached
    libraries. The "metastore" and "entitystore" options can now be
    set to a MemCache object or Memcached object:

    memcache = MemCache.new(['127.1.1.1', '127.1.1.2'], :namespace => "/foo")
    use Rack::Cache,
      :metastore => memcache,
      :entitystore => memcache

  * Fix "memcached://" metastore URL handling. The "memcached" variation
    blew up, the "memcache" version was fine.

## 0.5.0 / May 2009

  * Added meta and entity store implementations based on the
    memcache-client library. These are the default unless the memcached
    library has already been required.

  * The "allow_reload" and "allow_revalidate" options now default to
    false instead of true. This means we break with RFC 2616 out of
    the box but this is the expected configuration in a huge majority
    of gateway cache scenarios. See the docs on configuration
    options for more information on these options:
    http://tomayko.com/src/rack-cache/configuration

  * Added Google AppEngine memcache entity store and metastore
    implementations. To use GAE's memcache with rack-cache, set the
    "metastore" and "entitystore" options as follows:

        use Rack::Cache,
          :metastore   => 'gae://cache-meta',
          :entitystore => 'gae://cache-body'

    The 'cache-meta' and 'cache-body' parts are memcache namespace
    prefixes and should be set to different values.

## 0.4.0 / March 2009

  * Ruby 1.9.1 / Rack 1.0 compatible.

  * Invalidate cache entries that match the request URL on non-GET/HEAD

lib/rack/cache/headers.rb
require 'set'

module Rack::Cache
  # HTTP response header helper methods.
  module ResponseHeaders
    # Status codes of responses that MAY be stored by a cache or used in reply
    # to a subsequent request.
    #
    # http://tools.ietf.org/html/rfc2616#section-13.4
    CACHEABLE_RESPONSE_CODES = [
      200, # OK
      203, # Non-Authoritative Information
      300, # Multiple Choices
      301, # Moved Permanently
      302, # Found
      404, # Not Found
      410  # Gone
    ].to_set

    # A Hash of name=value pairs that correspond to the Cache-Control header.
    # Valueless parameters (e.g., must-revalidate, no-store) have a Hash value
    # of true. This method always returns a Hash, empty if no Cache-Control
    # header is present.
    def cache_control
      @cache_control ||= CacheControl.new(headers['Cache-Control'])
    end

    # Set the Cache-Control header to the values specified by the Hash. See
    # the #cache_control method for information on expected Hash structure.
    def cache_control=(value)
      if value.respond_to? :to_hash
        cache_control.clear
        cache_control.merge!(value)
        value = cache_control.to_s
      end

      if value.nil? || value.empty?
        headers.delete('Cache-Control')
      else
        headers['Cache-Control'] = value
      end
    end

    # Determine if the response is "fresh". Fresh responses may be served from
    # cache without any interaction with the origin. A response is considered
    # fresh when it includes a Cache-Control/max-age indicator or Expiration
    # header and the calculated age is less than the freshness lifetime.
    def fresh?
      ttl && ttl > 0
    end

    # Determine if the response is worth caching under any circumstance. Responses
    # marked "private" with an explicit Cache-Control directive are considered
    # uncacheable
    #
    # Responses with neither a freshness lifetime (Expires, max-age) nor cache
    # validator (Last-Modified, ETag) are considered uncacheable.
    def cacheable?
      return false unless CACHEABLE_RESPONSE_CODES.include?(status)
      return false if cache_control.no_store? || cache_control.private?
      validateable? || fresh?
    end

    # Determine if the response includes headers that can be used to validate
    # the response with the origin using a conditional GET request.
    def validateable?
      headers.key?('Last-Modified') || headers.key?('ETag')
    end

    # Mark the response "private", making it ineligible for serving other
    # clients.
    def private=(value)
      value = value ? true : nil
      self.cache_control = cache_control.
        merge('public' => !value, 'private' => value)
    end

    # Indicates that the cache must not serve a stale response in any
    # circumstance without first revalidating with the origin. When present,
    # the TTL of the response should not be overriden to be greater than the
    # value provided by the origin.
    def must_revalidate?
      cache_control.must_revalidate || cache_control.proxy_revalidate
    end

    # Mark the response stale by setting the Age header to be equal to the
    # maximum age of the response.
    def expire!
      headers['Age'] = max_age.to_s if fresh?
    end

    # The date, as specified by the Date header. When no Date header is present,
    # set the Date header to Time.now and return.
    def date
      if date = headers['Date']
        Time.httpdate(date)
      else
        headers['Date'] = now.httpdate unless headers.frozen?
        now
      end
    end

    # The age of the response.
    def age
      (headers['Age'] ||  [(now - date).to_i, 0].max).to_i
    end

    # The number of seconds after the time specified in the response's Date
    # header when the the response should no longer be considered fresh. First
    # check for a s-maxage directive, then a max-age directive, and then fall
    # back on an expires header; return nil when no maximum age can be
    # established.
    def max_age
      cache_control.shared_max_age ||
        cache_control.max_age ||
        (expires && (expires - date))
    end

    # The value of the Expires header as a Time object.
    def expires
      headers['Expires'] && Time.httpdate(headers['Expires'])
    end

    # The number of seconds after which the response should no longer
    # be considered fresh. Sets the Cache-Control max-age directive.
    def max_age=(value)
      self.cache_control = cache_control.merge('max-age' => value.to_s)
    end

    # Like #max_age= but sets the s-maxage directive, which applies only
    # to shared caches.
    def shared_max_age=(value)
      self.cache_control = cache_control.merge('s-maxage' => value.to_s)
    end

    # The response's time-to-live in seconds, or nil when no freshness
    # information is present in the response. When the responses #ttl
    # is <= 0, the response may not be served from cache without first
    # revalidating with the origin.
    def ttl
      max_age - age if max_age
    end

    # Set the response's time-to-live for shared caches to the specified number
    # of seconds. This adjusts the Cache-Control/s-maxage directive.
    def ttl=(seconds)
      self.shared_max_age = age + seconds
    end

    # Set the response's time-to-live for private/client caches. This adjusts
    # the Cache-Control/max-age directive.
    def client_ttl=(seconds)
      self.max_age = age + seconds
    end

    # The String value of the Last-Modified header exactly as it appears
    # in the response (i.e., no date parsing / conversion is performed).
    def last_modified
      headers['Last-Modified']
    end

    # The literal value of ETag HTTP header or nil if no ETag is specified.
    def etag
      headers['ETag']
    end

    # Determine if the response was last modified at the time provided.
    # time_value is the exact string provided in an origin response's
    # Last-Modified header.
    def last_modified_at?(time_value)
      time_value && last_modified == time_value
    end

    # Determine if response's ETag matches the etag value provided. Return
    # false when either value is nil.
    def etag_matches?(etag)
      etag && self.etag == etag
    end

    # Headers that MUST NOT be included with 304 Not Modified responses.
    #
    # http://tools.ietf.org/html/rfc2616#section-10.3.5
    NOT_MODIFIED_OMIT_HEADERS = %w[
      Allow
      Content-Encoding
      Content-Language
      Content-Length
      Content-MD5
      Content-Type
      Last-Modified
    ].to_set

    # Modify the response so that it conforms to the rules defined for
    # '304 Not Modified'. This sets the status, removes the body, and
    # discards any headers that MUST NOT be included in 304 responses.
    #
    # http://tools.ietf.org/html/rfc2616#section-10.3.5
    def not_modified!
      self.status = 304
      self.body = []
      NOT_MODIFIED_OMIT_HEADERS.each { |name| headers.delete(name) }
      nil
    end

    # The literal value of the Vary header, or nil when no header is present.
    def vary
      headers['Vary']
    end

    # Does the response include a Vary header?
    def vary?
      ! vary.nil?
    end

    # An array of header names given in the Vary header or an empty
    # array when no Vary header is present.
    def vary_header_names
      return [] unless vary = headers['Vary']
      vary.split(/[\s,]+/)
    end
  end

end
